### ISSUE TITLE
JOIN and WITH are not using the vector index

### ISSUE THREAD

**Issue by nadavrajuan** [pgvector/pgvector#174]
I am trying to query across three tables:
1. content
Includes id and embedding column.
(about 9M records)

2. user_content
Includes content_id and user_id - one to many linking table between user and content.
(about 18M records)

3. user_account
Some more metadata about the user that we would like to use to filter down users.
(about 600k records)

Ultimately when given a new content, we would like to find the most relevant users to this content. Out of the filtered down users list.

At this stage we have tried to get the most similar embedding records out of the content records of a subgroup of the users.

When we tried queries that use “join” or “with” to filter down the number of content records, we found out that the embedding index is not used, and the query takes about 15 minutes to complete.

When we split the queries: First get all content ids from user’s list, then we take all these ids (can be even 300-400k) and hardcode them in a new query that uses only the ‘content’ table of the embedding vector field (using WHERE IN)- postgres uses the index and the query time is 3s, which is great.
 
It’s super not scalable in running it query by query, and every other way we tried to handle having 2 tables in the query ended up by ignoring the embedding index which took the query to run in about 15 mins.
Have you come across this? Is there any suggestion or a solution you can think of?

**Comment by jkatz** [comment]
I need more information to understand what's happening:

1. What version of pgvector are you using? What version of PostgreSQL are you using?
2. How many dimensions are the vectors in the `embedding` column?
3. What are the indexes you have on all of the tables, and on which columns?
4. Do you have an example of the query that you are running? Do you also have an example `EXPLAIN` plan? How many probes are you using (i.e. `ivfflat.probes`).
5. How big of an instance are you using? What are some of the PostgreSQL performance parameters (`shared_buffers`, `effective_cache_size`, `work_mem`, `random_page_cost`, `seq_page_cost`)

**Comment by nadavrajuan** [comment]
1. Versions: 
- pgVector version: 0.4.1
- postgres engine version 15.3
- Postgres instance: db.t3.medium (2 vCPU, 4G)
2. Vector Dimensions: 768
3. Indexes:
**Content table index:** 


```
CREATE INDEX IF NOT EXISTS content_embedding_idx1
    ON content USING ivfflat
    (embedding vector_cosine_ops)
    TABLESPACE pg_default;
```
Also there is the primary key on this table.

**User_content index:**

```
CREATE INDEX IF NOT EXISTS inx_user_content_user_id
    ON superfy.user_content USING btree
    (user_id ASC NULLS LAST)
    TABLESPACE pg_default;
```

4. Example of a query:
```
with content_ids as (
	select content_id as id from user_content where user_id IN (
		733271,725866,693025… {LIST OF 895 USERS HERE}
	) limit 300000
)
select id, embedding <=> (select embedding from content where id = 2222222) as distance 
from content
where id in (select id from content_ids)
order by distance
limit 500;
```

Example of Explain

![explain](https://github.com/pgvector/pgvector/assets/1019717/bd1cbfd0-6421-4586-a2a4-9e25b1abae41)


- Ivfflat.probes We tried with 1 and with 10

5. PostgreSQL performance parameters:
- Shared_buffers "939960kB"
- Effective_cache_size "1879928kB"
- Work_mem "4MB"
- Random_page_cost 4
- Seq_page_cost 1

**Comment by ankane** [comment]
Hi @nadavrajuan, @jkatz may have some other ideas, but you could try doing the nearest neighbor search first, then the filtering (but could end up with less results than the final limit).

```sql
with similar_content as (
    select id, embedding <=> (select embedding from content where id = 2222222) as distance 
    from content
    order by distance
    limit 10000 -- adjust as needed
),
content_ids as (
     ...
)
select *
from similar_content
where id in (select id from content_ids)
order by distance
limit 500;
```

**Comment by jkatz** [comment]
Thanks for sharing the info @nadavrajuan. Just to note, I'm on the Amazon RDS team.

I agree with @ankane's assessment. I tried recreating a similar dataset to what you have and ran different variations of the query. I believe PostgreSQL is behaving correctly here: you've already selected a data set to search over in such a way that it wouldn't make sense to use the `ivfflat` index. Specifically, you've already filtered out the embeddings you want to search over in the CTE (`WITH` query), whereas `ivfflat` would help you select which rows to filter out. When I switched to a query similar to @ankane's, PostgreSQL chose the `ivfflat` index and had generally better performance results, though with the tradeoff of recall.

For the example that Andrew provided to work, you'll need to ensure that you set `ivfflat.probes` to a high enough value to ensure you return enough rows, assuming you're targeting the `LIMIT`.

Keep in mind the size of your data set as well when evaluating search performance. 9MM 768-dim vectors is ~27GB, and multiply this by 2 for the cost of the index. If your queries involve a lot of random access of your data, you will be swapping pages (which contain the data) in and out of memory frequently if your "cache" (`shared_buffers`) is too small.
